cpp/src/table/row_operators.cu

```diff
@@ -77,50 +159,77 @@ auto decompose_structs(table_view table,
                        host_span<order const> column_order         = {},
                        host_span<null_order const> null_precedence = {})
 {
+  auto sliced         = pushdown_struct_offsets(table);
+  auto linked_columns = detail::table_to_linked_columns(sliced);
+
   std::vector<column_view> verticalized_columns;
   std::vector<order> new_column_order;
   std::vector<null_order> new_null_precedence;
   std::vector<int> verticalized_col_depths;
-  for (size_type col_idx = 0; col_idx < table.num_columns(); ++col_idx) {
-    auto const& col = table.column(col_idx);
-    if (is_nested(col.type())) {
+  for (size_t col_idx = 0; col_idx < linked_columns.size(); ++col_idx) {
+    detail::linked_column_view const* col = linked_columns[col_idx].get();
+    if (is_nested(col->type())) {
       // convert and insert
       std::vector<column_view> r_verticalized_columns;
       std::vector<int> r_verticalized_col_depths;
-      std::vector<column_view> flattened;
+      std::vector<detail::linked_column_view const*> flattened;
       std::vector<int> depths;
       // TODO: Here I added a bogus leaf column at the beginning to help in the while loop below.
       //       Refactor the while loop so that it can handle the last case.
-      flattened.push_back(make_empty_column(type_id::INT32)->view());
-      std::function<void(column_view const&, int)> recursive_child = [&](column_view const& c,
-                                                                         int depth) {
-        flattened.push_back(c);
-        depths.push_back(depth);
-        if (c.type().id() == type_id::STRUCT) {
-          for (int child_idx = 0; child_idx < c.num_children(); ++child_idx) {
-            auto scol = structs_column_view(c);
-            recursive_child(scol.get_sliced_child(child_idx), depth + 1);
+      auto first_col =
+        std::make_shared<detail::linked_column_view>(make_empty_column(type_id::INT32)->view());
+      flattened.push_back(first_col.get());
+      std::function<void(detail::linked_column_view const*, int)> recursive_child =
+        [&](detail::linked_column_view const* c, int depth) {
+          flattened.push_back(c);
+          depths.push_back(depth);
+          if (c->type().id() == type_id::LIST) {
+            recursive_child(c->children[lists_column_view::child_column_index].get(), depth + 1);
+          } else if (c->type().id() == type_id::STRUCT) {
+            for (auto& child : c->children) {
+              recursive_child(child.get(), depth + 1);
+            }
           }
-        }
-      };
+        };
       recursive_child(col, 0);
       int curr_col_idx     = flattened.size() - 1;
-      column_view curr_col = flattened[curr_col_idx];
+      column_view curr_col = *flattened[curr_col_idx];
       while (curr_col_idx > 0) {
-        auto const& prev_col = flattened[curr_col_idx - 1];
-        if (not is_nested(prev_col.type())) {
+        auto const prev_col = flattened[curr_col_idx - 1];
```

Comment:
How about I rename `curr_col` to `temp_col`. `curr_col` would mean the column we are at in `linked_columns` but that's not what it is. It's a temporary column that's being built by traversing backwards over `linked_columns` and using `prev_col` to construct its parent.

Reply:
That would help too since you actually access `flattened[curr_col_idx]` again at one point to get the true current column. That still leaves the same confusion about `column_view` vs pointer to column_view though.

Reply:
> pointer to `column_view` though

There's no pointer to column_view in here. `prev_column` is a `linked_column_view *`. (BTW, it was changed from shared_ptr to raw). Should I rename to `pl_prev_col` (pointer to linked prev col)?

Reply:
Yup I saw the change from shared to raw pointers, thanks. I am mostly concerned about the fact that naively I would expect that `prev_col` and `curr_col` to be of the same type, so it might trip up future devs. But your suggestion makes me remember why I don't like Hungarian notation :joy: it usually doesn't make things any clearer. I'm fine with leaving this as is, I don't have a concrete suggestion for improvement here.